### ISSUE TITLE
Random 503 Service Unavailable from Load Balancer after upgrading service or manually restarting the LB container

### ISSUE THREAD

**Issue by rubenmch** [rancher/rancher#3360]
After upgrading a service that is a target of a load balancer service the load balancer randomly returns error 503 Service Unavailable.

It looks like the load balancer isn't updated after the service is upgraded and some connections go to the old container IP resulting in a 503 error and some connections go to the new container IP.

After stopping an starting again the load balancer service the random 503 errors stop happening.

This started happening after upgrading to rancher 0.51.0, it didn't happen in the version we had before.

Rancher Version:
Rancher	v0.51.0
Cattle	v0.130.0
User Interface	v0.78.0
Rancher Compose	v0.7.0

The load balancer and service containers are in the same host.

Steps to reproduce:

1. Create a Service
1. Create a Load Balancer Service and add the previous service as a target
1. Upgrade the service (Finishing the Upgrade doesn't make a difference, the error 503 is still returned randomly after finishing the upgrade)
1. Connect multiple times to the load balancer IP/port linked to the target service. Some connections will succeed and some will return error 503 Service Unavailable
1. Restart the load balancer service
1. Connect multiple times to the load balancer IP/port linked to the target service. All the connections will now succeed.

**Comment by deniseschannon** [comment]
Have you checked the haproxy.cfg to check if the configuration was updated after the upgrade of hte service was completed?

http://docs.rancher.com/rancher/faqs/troubleshooting/#how-can-i-see-the-configuration-of-my-load-balancer

**Comment by rubenmch** [comment]
@deniseschannon The haproxy.cfg config is up to date and has the correct IP of the new container.

While the load balancer is returning random 503 errors the following commands work in the load balancer container without errors:

* Running `curl http://NEWSERVICECONTAINERIP:PORT/` 100 times returns no errors
* `ping NEWSERVICECONTAINERIP` has no packets lost

The following doesn't solve the random 503 errors:

* Running `service haproxy reload` in the load balancer container
* Running `service haproxy restart` in the load balancer container

Only stopping the load balancer container and starting it again stops the random 503 errors.

Is there anything else I can check? I saw in another thread there aren't any haproxy logs in the current version of rancher right?

**Comment by lawsontyler** [comment]
I'm having the same issue.  I've enabled logging by doing the following inside the container:

`service rsyslog start`

then, modify /etc/haproxy/haproxy.cfg;

From:
```
global
        log 127.0.0.1 local0
        log 127.0.0.1 local1 notice
```

To:
```
global
        log /dev/log local0
        log /dev/log local1 notice
```

I'm still waiting for it to happen again so I can catch it in the logs.

**Comment by alena1108** [comment]
@rubenmch 

>> The following doesn't solve the random 503 errors:

Running service haproxy reload in the load balancer container
Running service haproxy restart in the load balancer container

That is not advisable - to reload haproxy by hand. Rancher reloads haproxy different way, and starting it with "service haproxy.." can lead into problems.

Most likely the issue you and @lawsontyler are seeing, is due to this bug:

https://github.com/rancher/rancher/issues/3355

https://forums.rancher.com/t/load-balancer-issue/1398/12

We are working on a fix right now.

**Comment by rubenmch** [comment]
@alena1108 @lawsontyler 

I enabled logging and got the following, entries containing `213 SC` are when the 503 error was returned:

```
Jan 21 17:12:20 0524d65c9119 haproxy[655]: 1.1.1.1:48847 [21/Jan/2016:17:1
2:08.054] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 9998/-1/121
40 213 SC 0/0/0/0/3 0/0                                                         
Jan 21 17:15:25 0524d65c9119 haproxy[902]: 1.1.1.1:39685 [21/Jan/2016:17:1
5:25.396] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/4/528 944
1 -- 1/1/0/1/0 0/0            
Jan 21 17:17:51 0524d65c9119 haproxy[655]: 1.1.1.1:64486 [21/Jan/2016:17:1
7:39.614] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10001/-1/12
148 213 SC 0/0/0/0/3 0/0                                                        
Jan 21 17:20:03 0524d65c9119 haproxy[655]: 1.1.1.1:6493 [21/Jan/2016:17:19
:51.116] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-b
ba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10000/-1/120
63 213 SC 0/0/0/0/3 0/0                                                         
Jan 21 17:21:12 0524d65c9119 haproxy[655]: 1.1.1.1:8916 [21/Jan/2016:17:21
:00.262] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-b
ba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10001/-1/121
12 213 SC 0/0/0/0/3 0/0                                                         
Jan 21 17:22:50 0524d65c9119 haproxy[655]: 1.1.1.1:34027 [21/Jan/2016:17:2
2:38.529] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 9999/-1/120
78 213 SC 0/0/0/0/3 0/0                                                         
Jan 21 17:22:54 0524d65c9119 haproxy[902]: 1.1.1.1:53024 [21/Jan/2016:17:2
2:54.400] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/2/206 944
1 -- 1/1/0/1/0 0/0                                                              
Jan 21 17:22:56 0524d65c9119 haproxy[902]: 1.1.1.1:53024 [21/Jan/2016:17:2
2:54.607] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/0/2242 94
41 -- 1/1/0/1/0 0/0                                                             
Jan 21 17:22:58 0524d65c9119 haproxy[902]: 1.1.1.1:53024 [21/Jan/2016:17:2
2:56.849] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/1271 94
41 -- 1/1/0/1/0 0/0                                                             
Jan 21 17:22:59 0524d65c9119 haproxy[902]: 1.1.1.1:53024 [21/Jan/2016:17:2
2:58.119] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/0/1636 94
41 -- 1/1/0/1/0 0/0                                                             
Jan 21 17:25:03 0524d65c9119 haproxy[655]: 1.1.1.1:41357 [21/Jan/2016:17:2
4:50.884] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10001/-1/12
143 213 SC 2/2/0/0/3 0/0                                                        
Jan 21 17:25:19 0524d65c9119 haproxy[655]: 1.1.1.1:62565 [21/Jan/2016:17:2
5:07.202] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 9999/-1/120
84 213 SC 1/1/0/0/3 0/0                                                         
Jan 21 17:25:34 0524d65c9119 haproxy[655]: 1.1.1.1:23800 [21/Jan/2016:17:2
5:22.349] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10001/-1/12
073 213 SC 1/1/0/0/3 0/0                                                        
Jan 21 17:25:47 0524d65c9119 haproxy[655]: 1.1.1.1:1403 [21/Jan/2016:17:25
:36.984] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-b
ba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 8370/-1/1043
9 213 SC 0/0/0/0/3 0/0                                                          
Jan 21 17:26:02 0524d65c9119 haproxy[655]: 1.1.1.1:32486 [21/Jan/2016:17:2
5:50.561] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/a742b970-5a74-487b-972c-5f8e448d8dde 10000/-1/12
062 213 SC 0/0/0/0/3 0/0                                                        
Jan 21 17:26:05 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:05.341] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/2/179 944
1 -- 1/1/0/1/0 0/0                                                              
Jan 21 17:26:07 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:05.521] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/1532 94
41 -- 1/1/0/1/0 0/0                                                             
Jan 21 17:26:07 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:07.052] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/873 944
1 -- 1/1/0/1/0 0/0                                                              
Jan 21 17:26:08 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:07.925] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/789 944
1 -- 1/1/0/0/0 0/0                                                              
Jan 21 17:26:09 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:08.713] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/826 944
1 -- 1/1/0/1/0 0/0                                                              
Jan 21 17:26:10 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:09.538] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/1102 94
41 -- 1/1/0/1/0 0/0                                                             
Jan 21 17:26:11 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:10.640] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/1/911 944
1 -- 1/1/0/1/0 0/0                                                              
Jan 21 17:26:12 0524d65c9119 haproxy[902]: 1.1.1.1:13945 [21/Jan/2016:17:2
6:11.551] bdc5c998-878b-49f5-bba7-68cacc843b5d_443_frontend~ bdc5c998-878b-49f5-
bba7-68cacc843b5d_443_4_backend/f71bc10f-e505-4e5c-906c-a4b959413ee2 0/0/814 944
1 -- 1/1/0/1/0 0/0                                                              
```

**Comment by rubenmch** [comment]
@alena1108 

Is there a recommended way to reload `/etc/haproxy/haproxy.cfg` from inside the container?

We need to enable tls v 1.0 removing `no-tlsv10` from haproxy.cfg and have so far been doing it by manually editing the haproxy.cfg file and then running `service haproxy reload`

**Comment by alena1108** [comment]
@rubenmch do you change haproxy.cfg inside the lb agent? Be aware of the fact that Rancher is going to override all the changes when it reloads the haproxy config (happens on every target instance restart, new link creation, etc). 

I would work on pushing this fix that can solve your problem: https://github.com/rancher/cattle/pull/1285

**Comment by rubenmch** [comment]
@alena1108 Yes, we change haproxy.cfg manually inside the lb agent and reload haproxy after every change to the instance, linked services or load balancer.

https://github.com/rancher/cattle/pull/1285 would solve this issue for us. It isn't included in Rancher 0.56.0 right?

**Comment by deniseschannon** [comment]
@rubenmch Correct, this is NOT in v0.56.0.

**Comment by alena1108** [comment]
Should be fixed with https://github.com/rancher/cattle/pull/1279

**Comment by sangeethah** [comment]
@alena1108: l will reopen this issue since this is dependent on the fix provided for #3355 which has been reopened.

**Comment by deniseschannon** [comment]
Verified in v0.63.0-rc1 with the following steps.

1. Create a Service
2. Create a Load Balancer Service and add the previous service as a target
3. Upgrade the service and finish upgrade
4. Connect multiple times to the load balancer IP/port linked to the target service. All the connections succeed.
